File: 0040-combination-sum-ii/0040-combination-sum-ii.java
class Solution {
    private HashSet<List<Integer>> res = new HashSet<>();
    private int[] nums;
    private int target;

    private void compute(List<Integer> path, int curr, int idx){
        if(curr == target){
            res.add(path);
            return;
        }

        if(idx == nums.length || curr>target){
            return;
        }

        List<Integer> temp = new ArrayList<>(path);
        temp.add(nums[idx]);
        curr += nums[idx];

        compute(temp, curr, idx+1);

        curr -= nums[idx];
        while (idx + 1 < nums.length && nums[idx] == nums[idx + 1]) {
            idx++;
        }
        compute(path, curr, idx+1);

    }

    public List<List<Integer>> combinationSum2(int[] candidates, int target) {
        this.nums = candidates;
        this.target = target;
        Arrays.sort(nums);
        List<Integer> path = new ArrayList<>();
        compute(path, 0, 0);
        return new ArrayList<>(res);
    }
}

File: 0040-combination-sum-ii/0040-combination-sum-ii.py
class Solution(object):
    def combinationSum2(self, candidates, target):
        candidates.sort()
        res = []

        def backtrack(start, path, total):
            if total == target:
                res.append(list(path))
                return
            if total > target:
                return

            for i in range(start, len(candidates)):
                if i > start and candidates[i] == candidates[i - 1]:
                    continue  # skip duplicates

                path.append(candidates[i])
                backtrack(i + 1, path, total + candidates[i])
                path.pop()

        backtrack(0, [], 0)
        return res
